tests/functional/instance_manager/test_instances.py

```diff
@@ -0,0 +1,192 @@
+
+        self.assertEqual(instance_manager.list(), [],
+                         instance_manager.list())
+        # "The initial instance file was not empty initially.")
+
+    def testServerStart(self):
+        """Started server writes itself to instance list"""
+
+        test_cfg = env.import_test_cfg(self._test_workspace)
+        codechecker_1 = test_cfg['codechecker_1']
+        start_server(codechecker_1, test_cfg, EVENT_1)
+
+        self.assertEqual(len(instance_manager.list()), 1,
+                         "The started server was not appended to the "
+                         "instance list.")
+
+        self.assertEqual(instance_manager.list()[0]['port'],
+                         codechecker_1['viewer_port'],
+                         "The written port in the instance list is not the "
+                         "one the server was started with!")
+
+        self.assertEqual(instance_manager.list()[0]['workspace'],
+                         self._test_workspace,
+                         "Server was brought up with invalid workspace.")
+
+    def testServerStartSecondary(self):
+        """Another started server appends itself to instance list"""
+
+        test_cfg = env.import_test_cfg(self._test_workspace)
+        codechecker_1 = test_cfg['codechecker_1']
+        codechecker_2 = test_cfg['codechecker_2']
+        start_server(codechecker_2, test_cfg, EVENT_2)
+
+        self.assertEqual(len(instance_manager.list()), 2,
+                         "The started server was not appended to the "
+                         "instance list.")
+
+        # Workspaces must match, servers were started in the same workspace.
+        instance_workspaces = [i['workspace'] for i in instance_manager.list()]
+        self.assertEqual(instance_workspaces[0], self._test_workspace,
+                         "Server was brought up with invalid workspace.")
+        self.assertEqual(instance_workspaces[0], instance_workspaces[1],
+                         "Server was brought up with invalid workspace.")
+
+        # Exactly one server should own each port generated
+        instance_ports = [i['port'] for i in instance_manager.list()]
+
+        self.assertTrue(any([True for p
+                             in instance_ports
+                             if p == codechecker_1['viewer_port']]),
+                        "Missing port for server started earlier!")
+
+        self.assertTrue(any([True for p
+                             in instance_ports
+                             if p == codechecker_2['viewer_port']]),
+                        "Missing port for server started later!")
+
+        self.assertEqual(len(set(instance_ports)), 2,
+                         "The servers started with the same port.")
+
+    def testShutdownRecordKeeping(self):
+        """Test that one server's shutdown keeps the other records."""
+
+        # NOTE: Do NOT rename this method. It MUST come lexicographically
+        # AFTER testServerStartSecondary
```

Comment:
If you look at the code, this method shuts down a server. To shut down a server, and to test that the shutdown actually does what it is expected to do, you need a server to be running before you attempt to shut it down.

And _nose_ executes tests in a lexicographical order.